[Arm64] Enable crossgen

sdmaclea opened the pull request:
Enable Arm64 Crossgen 

Fixes for arm64 #8998 

Depends on dotnet/core-setup#3987 ++

Works if I manually copy crossgen dependencies from core-setup

```bash
# Copy x64_arm64 crossgen
 cp -r \
    ../core-setup/packages/transport.runtime.linux-arm64.microsoft.netcore.runtime.coreclr/2.1.0-preview3-26404-06/tools/x64_arm64 \
    ./.nuget/packages/runtime.linux-arm64.microsoft.netcore.app/2.1.0-preview3-26404-01/tools

# Copy x64_arm64 libclrjit.so
cp -r \
    ../core-setup/packages/transport.runtime.linux-arm64.microsoft.netcore.jit/2.1.0-preview3-26404-06/runtimes/x64_arm64 \
    ./.nuget/packages/runtime.linux-arm64.microsoft.netcore.app/2.1.0-preview3-26404-01/runtimes/

```

@eerhardt 
Contains #8896 which I hope/expect will merge before dotnet/core-setup#3987 is ready

sdmaclea commented:
x86 debug was red around time of triggering...

test Windows_NT x86 Debug Build

sdmaclea commented:
@dotnet-bot test Windows_NT x86 Debug Build

sdmaclea commented:
I think this PR is probably dependent on getting new runtime bits with necessary tools i.e. #9015.

@johnbeisner Said in https://github.com/dotnet/cli/pull/8896#discussion_r177168919 that he set up VSO with triggers off.  The arm64 links are still not populated, so it looks like someone needs to establish triggers.

I would like to see this included in 2.1 branch, so that if and when we stabilize the arm64 runtime, we can use the cli.

johnbeisner commented:
@sdmaclea 
Triggers are now enabled: https://devdiv.visualstudio.com/DevDiv/Default/_build/index?context=allDefinitions&path=%5CDotNet%5CCLI%5Cmaster&definitionId=8817

sdmaclea commented:
@johnbeisner Thanks.  I can't see into your link so I don't exactly know what it means.  I am hoping it means that the arm64 linux is link in the readme will soon be populated.  https://dotnetcli.blob.core.windows.net/dotnet/Sdk/master/dotnet-sdk-latest-linux-arm64.tar.gz

sdmaclea commented:
Based on https://dotnet.myget.org/feed/dotnet-core/package/nuget/runtime.linux-arm64.Microsoft.NETCore.App

Needs 26407-01 or better.

kasper3 commented on src/redist/redist.csproj:
```diff
@@ -257,7 +257,7 @@
 
   <Target Name="CrossgenPublishDir"
-          Condition=" '$(DISABLE_CROSSGEN)' == '' And !$(Architecture.StartsWith('arm')) "
+          Condition=" '$(DISABLE_CROSSGEN)' == '' And !('$(Architecture)' == 'arm') "
```
`And '$(Architecture)' != 'arm' "`

sdmaclea commented:
I missed that the core-setup dependencies were updated 2 days ago.

Fixed nit above.  Rebased to pick up crosscomponent crossgen from core-setup.  Crossgen should work now.

@eerhardt Please trigger 'test ci please'

sdmaclea commented:
@dotnet-bot test Ubuntu16.04 x64 Debug Build 
@dotnet-bot test RHEL6 x64 Debug Build

eerhardt commented:
@dotnet-bot test ci please

sdmaclea commented:
Can this be merged?  I would be nice to include in 2.1

eerhardt commented:
@livarcocc @johnbeisner - thoughts on merging?

johnbeisner commented:
@eerhardt 
Is there a reason that we not enable crossgen for Linux arm as well?

eerhardt commented:
> Is there a reason that we are not enabling crossgen for Linux arm as well?

Not that I know of.  I think @sdmaclea's main concern is `linux-arm64` support. But if someone else wants to do `linux-arm`, that would be great.

sdmaclea commented:
linux-arm is slightly more complicated.  The current linux-arm crosscomponent need to run on x86 not x64.  @echesakovMSFT was working on enabling x64_arm crosscomponents.

eerhardt commented:
We do have a docker container that supports `linux-arm` crossgen.  We are using it in the core-setup build.

See https://github.com/dotnet/core-setup/pull/3892/files for the docker tag.

sdmaclea commented:
It was just too complicated for me to figure out how to enable linux-arm.  Enabling for linux-arm64 was trivial.

johnbeisner commented:
@sdmaclea 
Please rebase your 'PR-ARm64-CROSSGEN' branch to dotnet:release/2.1.3xx
Then we can retarget this PR to dotnet:release/2.1.3xx & I'll submit request for approval.

sdmaclea commented:
@johnbeisner This depends on coresetup 26407-01 or better (#9044 on master) which has not yet merged into dotnet/cli:release/2.1.3xx.  

Perhaps this should merge on the master branch first?

Let me know how you want me to proceed.

sdmaclea commented:
Rebased to the current master tip for now pending instructions.

sdmaclea commented:
Looks like this can be rebased on 2.1.3xx after #9076 merges

johnbeisner commented:
@sdmaclea 
"Looks like this can be rebased on 2.1.3xx after #9076 merges" - that seems like a reasonable plan.

johnbeisner commented:
@sdmaclea 
#9076 has merged; please move forward with the rebase/retarget

sdmaclea commented:
It is already done

sdmaclea commented:
> 2.2 is were most of the support for this will come anyways

"This" ?  linux-arm64?

> Could you please re-target to master (again, sorry) and then we can merge this.

Done

sdmaclea commented:
test Windows_NT_ES x64 Debug Build
